### PR TITLE
fix: correct the positional encoding of Transformer in pytorch examples

### LINKED ISSUE #19138: Wrong PositionalEncoding in the Transformer example
### Bug description

Hi,

I think there are several mistakes in the implemention of the PositionalEncoding in the [lightning/pytorch/demos/transformer.py.](https://github.com/Lightning-AI/pytorch-lightning/blob/master/examples/fabric/language_model/train.py)

Since the transformer is set to
https://github.com/Lightning-AI/pytorch-lightning/blob/275822d5aff848b7f0a4298498e905457b8455c9/src/lightning/pytorch/demos/transformer.py#L47
1. The shape of the pe should be [batch_size, len, dim]. This transpose operation is redundant.
https://github.com/Lightning-AI/pytorch-lightning/blob/275822d5aff848b7f0a4298498e905457b8455c9/src/lightning/pytorch/demos/transformer.py#L97
2. And the shape of `self.pe` should be `self.pe[:, :x.size(1)]`
https://github.com/Lightning-AI/pytorch-lightning/blob/275822d5aff848b7f0a4298498e905457b8455c9/src/lightning/pytorch/demos/transformer.py#L88
4. The result of the addition above is not assigned to x, so the pe will not take effect.  

Would you please let me know if I'm wrong?

### What version are you seeing the problem on?

master

### How to reproduce the bug

_No response_

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

_No response_

cc @borda

### PR DESCRIPTION
**Pull request by Galaxy-Husky**
## What does this PR do?

Fixes #19138

<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--20203.org.readthedocs.build/en/20203/

### CODE REVIEW

**Comment by Galaxy-Husky** [comment]
The difference between this example and the original one [pytorch example](https://github.com/pytorch/examples/blob/main/word_language_model/model.py) is that the default value of argument  `batch_first` is set to `False` in PyTorch, whereas here it is set to `True`.

In order to use the positional encoding correctly, I made the above changes. But I don't know how to write a test case.

**Comment by Galaxy-Husky** [comment]
@Borda Hi, could you tell me how to deal with the two failed workflows?

**Review by lantiga — APPROVED** [review]
Nice catch!
